### PR TITLE
Small fixes for better channels last performance

### LINKED ISSUE #1653: Tensor.scatter_ accepts bogus input where index size > src size
e.g.:
```
>>> torch.zeros(3,4).scatter_(0,torch.LongTensor([[0,0,0,0],[1,1,1,1],[2,2,2,2]]),torch.arange(1,9).resize_(2,4))

 1.0000e+00  2.0000e+00  3.0000e+00  4.0000e+00
 5.0000e+00  6.0000e+00  7.0000e+00  8.0000e+00
 2.7716e+20  7.2128e+22  4.4653e+30  7.2708e+31
[torch.FloatTensor of size 3x4]
```

valgrind reports reading off the end of the tensor:
```
==2959389== Invalid read of size 4
==2959389==    at 0x210EE636: THFloatTensor_scatter (in /data/users/gchanan/pytorch6/torch/lib/libTH.so.1)
==2959389==    by 0x169365A8: THPFloatTensor_scatter_(_object*, _object*, _object*) (TensorMethods.cpp:3493)
==2959389==    by 0x4F00901: _PyCFunction_FastCallDict (methodobject.c:231)
==2959389==    by 0x4F85F4B: call_function (ceval.c:4788)
==2959389==    by 0x4F88BBC: _PyEval_EvalFrameDefault (ceval.c:3275)
==2959389==    by 0x4F844BF: _PyEval_EvalCodeWithName (ceval.c:4119)
==2959389==    by 0x4F84942: PyEval_EvalCodeEx (ceval.c:4140)
==2959389==    by 0x4F8498A: PyEval_EvalCode (ceval.c:695)
==2959389==    by 0x4FB9345: run_mod (pythonrun.c:980)
==2959389==    by 0x4FB9345: PyRun_InteractiveOneObject (pythonrun.c:233)
==2959389==    by 0x4FB96AD: PyRun_InteractiveLoopFlags (pythonrun.c:112)
==2959389==    by 0x4FB97EB: PyRun_AnyFileExFlags (pythonrun.c:74)
==2959389==    by 0x4FD3BA8: run_file (main.c:320)
==2959389==    by 0x4FD3BA8: Py_Main (main.c:781)
==2959389==  Address 0x12f19840 is 0 bytes after a block of size 32 alloc'd
==2959389==    at 0x4C26B0F: malloc (in /usr/local/fbcode/gcc-5-glibc-2.23/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==2959389==    by 0x2106CC79: THAlloc (in /data/users/gchanan/pytorch6/torch/lib/libTH.so.1)
==2959389==    by 0x2106E037: THFloatStorage_resize (in /data/users/gchanan/pytorch6/torch/lib/libTH.so.1)
==2959389==    by 0x21087194: THFloatTensor_resize4d (in /data/users/gchanan/pytorch6/torch/lib/libTH.so.1)
==2959389==    by 0x2110DEF6: THFloatTensor_range (in /data/users/gchanan/pytorch6/torch/lib/libTH.so.1)
==2959389==    by 0x16981F53: THPFloatTensor_stateless_arange(_object*, _object*, _object*) (TensorMethods.cpp:3428)
==2959389==    by 0x4F00CD8: PyCFunction_Call (methodobject.c:98)
==2959389==    by 0x4EA9CE5: PyObject_Call (abstract.c:2246)
==2959389==    by 0x167EEAB1: THPUtils_dispatchStateless(_object*, char const*, _object*, _object*) (utils.cpp:154)
==2959389==    by 0x167C221F: THPModule_arange(_object*, _object*, _object*) (Module.cpp:300)
==2959389==    by 0x4F00901: _PyCFunction_FastCallDict (methodobject.c:231)
==2959389==    by 0x4F85F4B: call_function (ceval.c:4788)
==2959389==
```

### LINKED ISSUE #93536: Indeterministic code generation (triton backend)
Problem: Running same repro twice generates different code.
Expectation: The generated code is the same.

Repro
```python

from math import inf
import torch
from torch import tensor, device
import torch.fx as fx
import torchdynamo
from torchdynamo.testing import rand_strided
from torchdynamo.debug_utils import run_fwd_maybe_bwd
from torchdynamo.debug_utils import same_two_models

args = [((2, 3, 224, 224), (150528, 50176, 224, 1), torch.float32, 'cuda', False)]
args = [rand_strided(sh, st, dt, dev).requires_grad_(rg) for (sh, st, dt, dev, rg) in args]


from torch.nn import *
class Repro(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.self_self_stem_proj = Conv2d(3, 384, kernel_size=(16, 16), stride=(16, 16))
        self.self_self_stem_norm = Identity()
        self.self_self_blocks_0__norm1 = LayerNorm((384,), eps=1e-06, elementwise_affine=True)
        self.self_self_blocks_0__mlp_tokens_fc1 = Linear(in_features=196, out_features=384, bias=True)
        self.self_self_blocks_0__mlp_tokens_act = SiLU()
        self.self_self_blocks_0__mlp_tokens_drop1 = Dropout(p=0.0, inplace=False)
        self.self_self_blocks_0__mlp_tokens_fc2 = Linear(in_features=192, out_features=196, bias=True)
        self.self_self_blocks_0__mlp_tokens_drop2 = Dropout(p=0.0, inplace=False)
        self.self_self_blocks_0__drop_path = Identity()
        self.self_self_blocks_0__norm2 = LayerNorm((384,), eps=1e-06, elementwise_affine=True)
        self.self_self_blocks_0__mlp_channels_fc1 = Linear(in_features=384, out_features=1536, bias=True)
        self.self_self_blocks_0__mlp_channels_act = SiLU()
        self.self_self_blocks_0__mlp_channels_drop1 = Dropout(p=0.0, inplace=False)
        self.self_self_blocks_0__mlp_channels_fc2 = Linear(in_features=768, out_features=384, bias=True)
        self.self_self_blocks_0__mlp_channels_drop2 = Dropout(p=0.0, inplace=False)
        self.self_self_norm = LayerNorm((384,), eps=1e-06, elementwise_affine=True)
        self.self_self_head = Linear(in_features=384, out_features=1000, bias=True)



    def forward(self, x : torch.Tensor):
        self_self_stem_proj = self.self_self_stem_proj(x);  x = None
        flatten = self_self_stem_proj.flatten(2);  self_self_stem_proj = None
        transpose = flatten.transpose(1, 2);  flatten = None
        self_self_stem_norm = self.self_self_stem_norm(transpose);  transpose = None
        self_self_blocks_0__norm1 = self.self_self_blocks_0__norm1(self_self_stem_norm)
        transpose_1 = self_self_blocks_0__norm1.transpose(1, 2);  self_self_blocks_0__norm1 = None
        self_self_blocks_0__mlp_tokens_fc1 = self.self_self_blocks_0__mlp_tokens_fc1(transpose_1);  transpose_1 = None
        chunk = self_self_blocks_0__mlp_tokens_fc1.chunk(2, dim = -1);  self_self_blocks_0__mlp_tokens_fc1 = None
        getitem = chunk[0]
        getitem_1 = chunk[1];  chunk = None
        self_self_blocks_0__mlp_tokens_act = self.self_self_blocks_0__mlp_tokens_act(getitem_1);  getitem_1 = None
        mul = getitem * self_self_blocks_0__mlp_tokens_act;  getitem = self_self_blocks_0__mlp_tokens_act = None
        self_self_blocks_0__mlp_tokens_drop1 = self.self_self_blocks_0__mlp_tokens_drop1(mul);  mul = None
        self_self_blocks_0__mlp_tokens_fc2 = self.self_self_blocks_0__mlp_tokens_fc2(self_self_blocks_0__mlp_tokens_drop1);  self_self_blocks_0__mlp_tokens_drop1 = None
        self_self_blocks_0__mlp_tokens_drop2 = self.self_self_blocks_0__mlp_tokens_drop2(self_self_blocks_0__mlp_tokens_fc2);  self_self_blocks_0__mlp_tokens_fc2 = None
        transpose_2 = self_self_blocks_0__mlp_tokens_drop2.transpose(1, 2);  self_self_blocks_0__mlp_tokens_drop2 = None
        self_self_blocks_0__drop_path = self.self_self_blocks_0__drop_path(transpose_2);  transpose_2 = None
        add = self_self_stem_norm + self_self_blocks_0__drop_path;  self_self_stem_norm = self_self_blocks_0__drop_path = None
        self_self_blocks_0__norm2 = self.self_self_blocks_0__norm2(add)
        self_self_blocks_0__mlp_channels_fc1 = self.self_self_blocks_0__mlp_channels_fc1(self_self_blocks_0__norm2);  self_self_blocks_0__norm2 = None
        chunk_1 = self_self_blocks_0__mlp_channels_fc1.chunk(2, dim = -1);  self_self_blocks_0__mlp_channels_fc1 = None
        getitem_2 = chunk_1[0]
        getitem_3 = chunk_1[1];  chunk_1 = None
        self_self_blocks_0__mlp_channels_act = self.self_self_blocks_0__mlp_channels_act(getitem_3);  getitem_3 = None
        mul_1 = getitem_2 * self_self_blocks_0__mlp_channels_act;  getitem_2 = self_self_blocks_0__mlp_channels_act = None
        self_self_blocks_0__mlp_channels_drop1 = self.self_self_blocks_0__mlp_channels_drop1(mul_1);  mul_1 = None
        self_self_blocks_0__mlp_channels_fc2 = self.self_self_blocks_0__mlp_channels_fc2(self_self_blocks_0__mlp_channels_drop1);  self_self_blocks_0__mlp_channels_drop1 = None
        self_self_blocks_0__mlp_channels_drop2 = self.self_self_blocks_0__mlp_channels_drop2(self_self_blocks_0__mlp_channels_fc2);  self_self_blocks_0__mlp_channels_fc2 = None
        self_self_blocks_0__drop_path_1 = self.self_self_blocks_0__drop_path(self_self_blocks_0__mlp_channels_drop2);  self_self_blocks_0__mlp_channels_drop2 = None
        add_1 = add + self_self_blocks_0__drop_path_1;  add = self_self_blocks_0__drop_path_1 = None
        self_self_norm = self.self_self_norm(add_1);  add_1 = None
        mean = self_self_norm.mean(dim = 1);  self_self_norm = None
        self_self_head = self.self_self_head(mean);  mean = None
        return (self_self_head,)



mod = Repro().cuda()
opt_mod = torchdynamo.optimize("inductor")(mod)


mod.eval()
opt_mod.eval()
with torch.cuda.amp.autocast(enabled=False):
    assert same_two_models(mod, mod, args), "Eager itself failed"
    assert same_two_models(mod, opt_mod, args), "Dynamo failed"
```

Run the repro with
```bash
TORCHINDUCTOR_TRACE=1 python repro_gmixer_24_224_min.py
```
Check the forward graph 0.

cc @ezyang @soumith @msaroufim @wconstab @ngimel @bdhirsh

### PR DESCRIPTION
**Pull request by ngimel**
1) don't codegen maxpool backward, it's exceedingly slow
2) better determine reduction variables for more accurate hints
3) deterministic iteration order for reduction arguments, take into account all full size reduction argument, for hints break ties to outer reduction

fixes pytorch/pytorch#93536

cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

### CODE REVIEW

File: torch/_inductor/ir.py [hunk 1]
```diff
@@ -625,20 +625,24 @@ def outer_reduction_splits(reduction_numel_hint, numel_hint):
...
-        for md in read_writes.reads:
+        for md in sorted(read_writes.reads):
             if all([r in md.index.free_symbols for r in range_vars]):
                 index = md.index
                 break
         if not index:
             # TODO determine splits when all inputs are broadcasted
             return ReductionHint.DEFAULT, 1
-        for md in read_writes.writes:
+        for md in sorted(read_writes.writes):
```
Comment (jgong5) [inline torch/_inductor/ir.py:635]: Not sure if this change would fix the issue: https://github.com/pytorch/torchdynamo/issues/1653

Comment (ngimel) [reply torch/_inductor/ir.py:635]: Yes it will

**Review by jansel — APPROVED** [review]

File: torch/_inductor/ir.py [hunk 2]
```diff
@@ -748,6 +770,7 @@ def create(
         reduction_hint: ReductionHint = ReductionHint.DEFAULT,
     ):
         reduction_numel = V.graph.sizevars.simplify(sympy_product(reduction_ranges))
+        xnumel = V.graph.sizevars.simplify(sympy_product(ranges))
```
Comment (Chillee) [inline torch/_inductor/ir.py:773]: Is this still needed?

**Review by Chillee — APPROVED** [review]

File: torch/_inductor/ir.py (hunk 2 above)
Comment (ngimel) [reply torch/_inductor/ir.py:773]: Looks like it isn't

Comment (ngimel) [inline torch/_inductor/ir.py:773]:
```suggestion
```

**Pull request closed without merging**